### PR TITLE
[cryptolib] Add Ibex drivers for P-384 ECDH and ECDSA on OTBN

### PR DESCRIPTION
**Pull request by wettermo**
This PR 

- adds Ibex driver code for P-384 ECC on OTBN. 
  -> In case of ECDSA the code is split up into individual files for keygen, sign, and verify. This is because each function is a separate OTBN application. But still, everything is put together in a single ecdsa_p384 file, which then is used by the top level ecc driver file.
  -> Similar, curve point validation is put into an own driver file, as it also is a separate OTBN application. It is executed before ECDH shared key generation and ECDSA signature verification, to check if the public key is valid (= is a curve point).

- adds the corresponding functests for P-384 ECDH and ECDSA.
- fixes some memory issues in the P-384 top level OTBN binaries, which we discovered during development.

### CODE REVIEW

File: sw/device/lib/crypto/impl/ecc/ecdh_p384.c
Comment (jadephilipoom) [inline sw/device/lib/crypto/impl/ecc/ecdh_p384.c:56]: nit: maybe these can go in an anonymous enum to match the other files?

File: sw/device/lib/crypto/impl/ecc/ecdh_p384.h
Comment (jadephilipoom) [inline sw/device/lib/crypto/impl/ecc/ecdh_p384.h:36]: Add this macro to ensure that callers get warnings if they don't check the error code:
```suggestion
OT_WARN_UNUSED_RESULT
status_t ecdh_p384_keypair_start(void);
```
Same goes for every other function declaration added in this PR that returns `status_t`. Even internal functions declared with `static` should get this macro to help protect against callers not forwarding failures.

File: sw/device/lib/crypto/impl/ecc/ecdsa_p384_verify.c
Comment (jadephilipoom) [inline sw/device/lib/crypto/impl/ecc/ecdsa_p384_verify.c:126]: I see that this helper function is repeated for both `sign` and `verify`; Maybe this could go in `p384_common`, and take the OTBN destination address as an argument? Alternatively, since it only has one call site in each file, it could be inlined.

**Review by jadephilipoom — APPROVED** [review]
Looks good, a few comments here and there. The most important note in my view is the addition of `OTCRYPTO_WARN_IF_UNUSED_RESULT`.

As a side note for the future, no need to change for this PR: it would be nice to submit changes like this in smaller chunks so they're easier to review. For example, in this case the operation-specific interfaces e.g. `ecdsa_p384_sign.c` and `ecdsa_p384_verify.c` could each have been submitted separately, then the `ecc.c` changes, then the test. I know for cryptolib it's tricky because you can't test it until you have the whole thing locally, and that often results in changes throughout the implementation. What I do is I implement the whole thing locally up to the test, get the test passing, and then create new branches where I `git checkout` specific files from the full-implementation branch for smaller PRs. I usually try to keep individual PRs below 500 lines of code added if I can. It's more work for the implementer, but it's easier on reviewers and often results in quicker turnaround time for PRs :slightly_smiling_face:

**Comment by wettermo** [comment]
Thanks Jade for the detailed review!
Sorry for not splitting it up into more commits, my bad. You're absolutely right. Noted for the next PR!

File: sw/device/lib/crypto/impl/ecc/ecdh_p384.c
Comment (wettermo) [reply sw/device/lib/crypto/impl/ecc/ecdh_p384.c:56]: Good point. But then we should also change this in `ecdh_p256.c`. Because that's where I copied it from.

File: sw/device/lib/crypto/impl/ecc.c
Comment (wettermo) [reply sw/device/lib/crypto/impl/ecc.c:157]: You're right, `OTCRYPTO_NOT_IMPLEMENTED` fits better here.
Yes it's definitely still a TODO. The key side-load above is not intentional, it's just a remnant from copying the p256 code. Sorry, I wasn't very thorough here.

File: sw/device/lib/crypto/impl/ecc.c
Comment (wettermo) [reply sw/device/lib/crypto/impl/ecc.c:391]: Thanks, missed this.

File: sw/device/lib/crypto/impl/ecc.c
Comment (wettermo) [reply sw/device/lib/crypto/impl/ecc.c:389]: You're right! Also just another remnant from code copying...

File: sw/device/lib/crypto/impl/ecc.c
Comment (wettermo) [reply sw/device/lib/crypto/impl/ecc.c:158]: Hmm might be true. But the we also should change it in the condition a few lines above for p256, because that's where I copied it from.

File: sw/device/lib/crypto/impl/ecc.c
Comment (wettermo) [reply sw/device/lib/crypto/impl/ecc.c:627]: Ups, thanks for pointing it out. This isn't intended. I somehow mixed it up. I'll rename everything to `sig_p256/sig_p384`.

File: sw/device/lib/crypto/impl/ecc/ecdsa_p384.h
Comment (wettermo) [reply sw/device/lib/crypto/impl/ecc/ecdsa_p384.h:20]: > Is the purpose to provide one single header file that ecc.c can import, so the internal structure is hidden?
Exactly, that was the idea. But, you're right a `.c` file doesn't really make sense here. I'll remove it and add an explaining comment in the header file.

File: sw/device/lib/crypto/impl/ecc/ecdh_p384.h
Comment (wettermo) [reply sw/device/lib/crypto/impl/ecc/ecdh_p384.h:36]: Sure, makes sense. But, as far as I see, this then would also concern a lot of other files which aren't even affected by this PR.

Comment (jadephilipoom) [reply sw/device/lib/crypto/impl/ecc/ecdh_p384.h:36]: Then those files have errors, but I still don't want to merge new errors in when code is being added. I've been trying to fix these missing macros every time I see them, even if I'm just vaguely editing code in the same file.

**Comment by wettermo** [comment]
Since there are quite a few point that are also relevant beyond this PR, I would suggest that, for now, I only address everything which is directly relevant to this PR. For everything else, that for example affects p256 code, I would open a follow-up PR to add these changes. What do you think @jadephilipoom ?

**Comment by jadephilipoom** [comment]
> Since there are quite a few point that are also relevant beyond this PR, I would suggest that, for now, I only address everything which is directly relevant to this PR. For everything else, that for example affects p256 code, I would open a follow-up PR to add these changes. What do you think @jadephilipoom ?

I'm not certain which of these you mean, but my stance is: it's fine with me to not fix the corresponding errors in P-256 code here and do those in a follow-up PR. However, I would like to see the issues fixed for the new P-384 code in this PR before it's merged, even if the issues exist in P-256 as well -- follow-up PRs aren't guaranteed to happen, and even when the existing codebase isn't perfect I think it's good practice to try not to merge new flaws.

**Comment by wettermo** [comment]
I updated the code. Now everything pointed out in the comments above should be taken care of. If you maybe could do a quick double check again @jadephilipoom, that would be nice 🙂.
For points which also affect files that are not relevant for this PR, I will open an additional PR. (comment about `OT_WARN_UNUSED_RESULT` and anonymous enum)

**Comment by jadephilipoom** [comment]
Looks good to me, thanks! Happy to merge once it's passing CI.

**Pull request merged**
